### PR TITLE
[TS-620] Add phone, date, and address parsing

### PR DESCRIPTION
**Pull request by alexvbush**
## Issue Description

This PR adds support for date, phone, and address parsing in ActiveLabel library. This is needed to support all types of data parsing in WO description field.

## Code Changes

- added several UT cases for the new data parsing types
- followed the existing pattern of ActiveLabel public APIs and internal NSRegularExpression caching in a static dictionary

## Automated Test Cases

- PhoneParsingTests
- AddressParsingTests
- DateParsingTests

## Issue Tickets
TS-620

### CODE REVIEW

File: ActiveLabelTests/DateParsingTests.swift [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+//  Created by Alex Bush | Upkeep on 6/29/21.
+//  Copyright © 2021 Optonaut. All rights reserved.
+//
+
+import XCTest
+@testable import ActiveLabel
+
+final class DateParsingTests: BaseTestCase {
+    // MARK: - BEGIN Parsing
+    func testItShouldNotRecognizeAnyDatesInTextWithoutAddresses() {
```
Comment (SandeepAggarwal) [inline ActiveLabelTests/DateParsingTests.swift]: pls change the name: testItShouldNotRecognizeAnyDatesInTextWithoutDates

**Review by SandeepAggarwal — CHANGES_REQUESTED** [review]
Awesome!

**Review by SandeepAggarwal — APPROVED** [review]

File: ActiveLabel/RegexFactory.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .hashtag:
+            let createdRegex = try? NSRegularExpression(pattern: hashtagPattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .custom(pattern: let pattern):
+            let createdRegex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
```
Comment (SandeepAggarwal) [inline ActiveLabel/RegexFactory.swift]: we can refactor out these 2 repeating lines of code

Comment (SandeepAggarwal) [reply ActiveLabel/RegexFactory.swift]: what benefits a class instance of RegexParser provides over a static instance?

File: ActiveLabel/ActiveLabel.swift [hunk 3]
```diff
@@ -185,6 +185,10 @@ typealias ElementTuple = (range: NSRange, element: ActiveElement, type: ActiveTy
         didSet { textContainer.lineBreakMode = lineBreakMode }
     }
     
+    /// This is exposed as an internal var so that in UTs we can swap it for another implementation mock and stub it
+    var activeBuilder: ActiveBuilderInterface = ActiveBuilder(regexParser: RegexParser())
```
Comment (SandeepAggarwal) [inline ActiveLabel/ActiveLabel.swift]: why can't we inject it in a init method?

File: ActiveLabelTests/IntegrationTests/PhoneParsingIntegrationTests.swift [hunk 4]
```diff
@@ -9,7 +9,7 @@
 import XCTest
 @testable import ActiveLabel
 
-final class PhoneParsingTests: BaseTestCase {
+final class PhoneParsingIntegrationTests: BaseIntegrationTestCase {
```
Comment (SandeepAggarwal) [inline ActiveLabelTests/IntegrationTests/PhoneParsingIntegrationTests.swift:12]: BaseTestCase vs BaseIntegrationTestCase?

**Review by SandeepAggarwal — CHANGES_REQUESTED** [review]

File: ActiveLabel/RegexFactory.swift (hunk 2 above)
Comment (alexvbush) [reply ActiveLabel/RegexFactory.swift]: >we can refactor out these 2 repeating lines of code

👍

>what benefits a class instance of RegexParser provides over a static instance?

It's finally injectable and testable

File: ActiveLabel/ActiveLabel.swift (hunk 3 above)
Comment (alexvbush) [reply ActiveLabel/ActiveLabel.swift]: for backwards compatibility - existing clients of this code are using built-in inits from UIKit such as `init(frame: CGRect)` and `init?(coder aDecoder: NSCoder)`, etc. We can't break their code by requiring it in init. Plus it doesn't seem like this is something that users of this library should be able to configure. That's why this property is not `open` or `public` but instead `internal`

File: ActiveLabelTests/IntegrationTests/PhoneParsingIntegrationTests.swift (hunk 4 above)
Comment (alexvbush) [reply ActiveLabelTests/IntegrationTests/PhoneParsingIntegrationTests.swift:12]: BaseTestCase was actually settings things up for integration testing rather then unit testing, hence the name change.

**Review by SandeepAggarwal — APPROVED** [review]

**Pull request merged**